pkg/crawler/crawler.go

```diff
@@ -82,334 +128,474 @@ func NewCrawler(opt Option) (Crawler, error) {
+	var processedPrefixes int
+	for prefix, err := range l.client.TopLevelPrefixes(ctx) {
+		// Handle errors during iteration immediately
+		if err != nil {
+			return xerrors.Errorf("failed to list top-level prefixes: %w", err)
+		}
+
+		g.Go(func() error {
+			return l.processPrefix(ctx, prefix, itemCh)
+		})
+		processedPrefixes++
 	}
 
-	var foundVersions []types.Version
-	// Get versions from the DB (if exists) to reduce the number of requests to the server
-	savedVersion, err := c.versionsFromDB(groupID, artifactID)
-	if err != nil {
-		return xerrors.Errorf("unable to get list of versions from DB: %w", err)
+	// Wait for all prefix processing to complete
+	if err := g.Wait(); err != nil {
+		return xerrors.Errorf("error processing prefixes: %w", err)
 	}
-	// Check each version dir to find links to `*.jar.sha1` files.
-	for dirVersion, itemNames := range dirs {
-		var dirVersionSha1 []byte
-		var versions []types.Version
-
-		for _, itemName := range itemNames {
-			_, _, _, ver := parseItemName(itemName)
-			sha1, ok := savedVersion[ver]
-			if !ok {
-				sha1, err = c.fetchSHA1(ctx, itemName)
-				if err != nil {
-					return xerrors.Errorf("unable to fetch sha1: %s", err)
-				}
-			}
-			// Save sha1 for the file where the version is equal to the version from the directory name in order to remove duplicates later
-			// Avoid overwriting dirVersion when inserting versions into the database (sha1 is uniq blob)
-			// e.g. `cudf-0.14-cuda10-1.jar.sha1` should not overwrite `cudf-0.14.jar.sha1`
-			// https://repo.maven.apache.org/maven2/ai/rapids/cudf/0.14/
-			if ver == dirVersion {
-				dirVersionSha1 = sha1
-			} else {
-				versions = append(versions, types.Version{
-					Version: ver,
-					SHA1:    sha1,
-				})
-			}
+
+	l.logger.Info("GCS artifact listing completed", slog.Int("total", int(l.processed.Load())), slog.Int("prefixes", processedPrefixes))
+	return nil
+}
+
+// processPrefix processes a single prefix, listing all matching artifacts
+func (l *Lister) processPrefix(ctx context.Context, prefix string, itemCh chan<- string) error {
+	// Use the JARSHA1Files iterator from GCS
+	for item, err := range l.client.JARSHA1Files(ctx, prefix) {
+		// Handle errors during iteration immediately
+		if err != nil {
+			return xerrors.Errorf("error listing JAR SHA1 files for prefix %s: %w", prefix, err)
 		}
-		// Remove duplicates of dirVersionSha1
-		versions = lo.Filter(versions, func(v types.Version, _ int) bool {
-			return !bytes.Equal(v.SHA1, dirVersionSha1)
-		})
 
-		if dirVersionSha1 != nil {
-			versions = append(versions, types.Version{
-				Version: dirVersion,
-				SHA1:    dirVersionSha1,
-			})
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case itemCh <- item:
+			// Thread-safe increment using atomic
+			processed := l.processed.Add(1)
+
+			// Log every 100,000 processed items
+			if processed%100000 == 0 {
+				l.logger.Info(fmt.Sprintf("Listed %d artifacts", processed))
+			}
```

Comment:
You check items only for 1 prefix here.
And you will see duplicates (e.g. 100000 items from `com` and from `org` prefixes).

I am not wrong - you can move this logic to:
https://github.com/aquasecurity/trivy-java-db/blob/9fdf0b0c524d465b1a0be317babdadad2377d70f/pkg/crawler/crawler.go#L405

Reply:
I guess your confusion comes from a behavior of atomic.Uint32.
https://go.dev/play/p/q2eIYgCPvbf

The return value is just to preserve the current value because l.processed is incremented by some goroutines and the value may skip 100,000 when it's loaded.

Reply:
Added a comment
https://github.com/aquasecurity/trivy-java-db/pull/57/commits/03f74a1fceacd0acc3a187c26d829f15ceaf32b4

Reply:
hm... it looks like I got something mixed up.
I thought we were creating a separate `processed` instance for each goroutine.

Now i see that there is no problem with this.